uqcsbot/dominos_coupons.py

```diff
@@ -183,17 +158,62 @@ def keyword_matches(self, keyword: str) -> bool:
 
 
 def _get_coupons(
-    n: int, ignore_expiry: bool, keywords: List[str], source: str
-) -> List[Coupon]:
+    n: int,
+    ignore_expiry: bool,
+    keywords: List[str],
+    source: Literal["couponese", "frugalfeeds", "both"],
```

Comment:
Given that you use `Literal["couponese", "frugalfeeds", "both"]`, it might be best to create a new type (see [`advent.py`](https://github.com/UQComputingSociety/uqcsbot-discord/blob/main/uqcsbot/advent.py#L45) for an example).

Reply:
I am fairly unfamiliar with creating new types, but I have had a crack at implementing this. I'm unsure if I'm following best practices here:

`CouponSource = Literal[
    "frugalfeeds", 
    "couponese", 
    "both"]`

I've declared this globally and have replaced all instances of `Literal["couponese", "frugalfeeds", "both"]` declaration with CouponSource, e.g. 

`source: CouponSource = "both"`, 
`def _get_coupons(
    n: int,
    ignore_expiry: bool,
    keywords: List[str],
    source: CouponSource,
) -> Tuple[List[Coupon], List[str]]:`,

And then have used `source` as I was using it before, e.g.. 

`coupons, failed_urls = _get_coupons(
            number_of_coupons, ignore_expiry, keywords.split(), source
        )`

`if source == "both":
            description_string = f"Sourced from [FrugalFeeds]({FRUGAL_FEEDS_DOMINOS_URL}) and [Couponese]({COUPONESE_DOMINOS_URL})"`

Am I on the right path here? Additionally, I spotted another possible area of improvement - I've used `Literal["couponese", "frugalfeeds"]]` a couple of times. Should I declare it as its own type as well? Is there a better way to approach this? It's just `CouponSource` without both.

Reply:
That's exactly it. It makes the code easier to read and reduces repetition. The `SingleSource` type is also a goo idea. Personally, I would have made `CouponSource = SingleSource | Literal["both"]`, but how you have implemented it is also very clean IMO.